src/circuit/gadget/sinsemilla/chip/hash_to_point.rs

```diff
@@ -0,0 +1,446 @@
+            pallas::Affine,
+            { sinsemilla::K },
+            { sinsemilla::C },
+        >>::Message,
+    ) -> Result<(EccPoint, Vec<Vec<CellValue<pallas::Base>>>), Error> {
+        let config = self.config().clone();
+        let mut offset = 0;
+
+        // Get the `x`- and `y`-coordinates of the starting `Q` base.
+        let x_q = *Q.coordinates().unwrap().x();
+        let y_q = *Q.coordinates().unwrap().y();
+
+        // Initialize the accumulator to `Q`.
+        let (mut x_a, mut y_a): (X<pallas::Base>, Y<pallas::Base>) = {
+            // Constrain the initial x_q to equal the x-coordinate of the domain's `Q`.
+            let fixed_x_q =
+                region.assign_fixed(|| "fixed x_q", config.constants, offset, || Ok(x_q))?;
+            let x_q_cell = region.assign_advice(|| "x_q", config.x_a, offset, || Ok(x_q))?;
+            region.constrain_equal(&config.perm, fixed_x_q, x_q_cell)?;
+
+            // This cell gets copied into itself by the first call to `hash_piece` below.
+            let x_a = CellValue::new(x_q_cell, Some(x_q));
+
+            // Constrain the initial x_a, lambda_1, lambda_2, x_p using the fixed y_q
+            // initializer.
+            region.assign_fixed(|| "fixed y_q", config.fixed_y_q, offset, || Ok(y_q))?;
+
+            let y_a = Some(y_q);
+
+            (x_a.into(), y_a.into())
+        };
+
+        let mut zs_sum: Vec<Vec<CellValue<pallas::Base>>> = Vec::new();
+
+        // Hash each piece in the message except the final piece.
+        for piece in message[0..(message.len() - 1)].iter() {
+            // The value of the accumulator after this piece is processed.
+            let (x, y, zs) = self.hash_piece(region, offset, piece, x_a, y_a)?;
+
+            // Since each message word takes one row to process, we increase
+            // the offset by `piece.num_words` on each iteration.
+            offset += piece.num_words();
+
+            // Update the accumulator to the latest value.
+            x_a = x;
+            y_a = y;
+            zs_sum.push(zs);
+        }
+
+        // Hash the final message piece.
+        let y_a = {
+            let piece = &message[message.len() - 1];
+            // The value of the accumulator after this piece is processed.
+            let (x, y, zs) = self.hash_piece(region, offset, piece, x_a, y_a)?;
+
+            // Since each message word takes one row to process, we increase
+            // the offset by `piece.num_words` on each iteration.
+            offset += piece.num_words();
+
+            // Update the accumulator to the latest value.
+            x_a = x;
+            y_a = y;
+            zs_sum.push(zs);
+
+            // Assign and constrain the final `y_a`.
+            region.assign_fixed(
+                || "qs_2 = 2 on final row",
+                config.q_sinsemilla2,
+                offset - 1,
+                || Ok(pallas::Base::from_u64(2)),
+            )?;
```

Comment:
This causes a double-assignment, as it is (correctly) overwriting the final `0` assigned inside `hash_piece`:
![image](https://user-images.githubusercontent.com/4993799/122651540-7a46e000-d131-11eb-8764-52c432361997.png)

We should pass a `last` boolean flag into `hash_piece` (which we can detect in the main loop), and then set `qs_2` for the last piece to either `0` or `2` based on that flag.
```suggestion
            // Assign and constrain the final `y_a`.
```

Reply:
What's the extra cell in the fourth row that is the only one in the sixth column?

Reply:
That's `fixed_y_A`. The first three columns in the screenshot are the lookup table, and the other two are `q_sinsemilla1` (a `Selector`, which now has a darker blue background) and `q_sinsemilla2`.

The impl isn't setting `fixed_y_A` to zero on every other row because we aren't requiring it. `MockProver` can't detect that this gate is active because it doesn't involve a selector, and `MockProver` can't make any assumptions about other gates (which could be arbitrarily more complex).

Reply:
I think that the case of "any non-zero value means selected" is common enough to support in `MockProver`.

Reply:
Opened up https://github.com/zcash/halo2/issues/301 to look into this.